giftless/auth/github.py

```diff
@@ -238,67 +280,226 @@ def from_dict(cls, data: Mapping[str, Any]) -> "Config":
     """
 
-    id: str
-    github_id: str
+    # authenticator config
+    cfg: Config
+    # original flask request to be authenticated
+    request: dataclasses.InitVar[flask.Request]
+    # fields inferred from request
+    org: str = dataclasses.field(init=False)
+    repo: str = dataclasses.field(init=False)
+    user: str | None = dataclasses.field(init=False)
+    token: str = dataclasses.field(init=False)
+    # GitHub api call variables
+    _api_url: str = dataclasses.field(init=False)
+    _api_headers: dict[str, str] = dataclasses.field(
+        init=False,
+        default_factory=lambda: {"Accept": "application/vnd.github+json"},
+    )
+    # requests session to reuse a connection to GitHub
+    _session: requests.Session | None = dataclasses.field(
+        init=False, default=None
+    )
+    _exit_stack: ExitStack = dataclasses.field(
+        init=False, default_factory=ExitStack
+    )
 
-    @classmethod
-    def from_token(
-        cls, token_data: Mapping[str, Any]
-    ) -> "_CoreGithubIdentity":
-        return cls(*itemgetter("login", "id")(token_data))
+    def __post_init__(self, request: flask.Request) -> None:
+        org_repo_getter = itemgetter("organization", "repo")
+        self.org, self.repo = org_repo_getter(request.view_args or {})
+        self.user, self.token = self._extract_auth(request)
 
+        self._api_url = self.cfg.api_url
+        self._api_headers["Authorization"] = f"Bearer {self.token}"
+        if self.cfg.api_version:
+            self._api_headers["X-GitHub-Api-Version"] = self.cfg.api_version
 
-class GithubIdentity(Identity):
-    """User identity belonging to an authentication token.
+    def __enter__(self) -> "CallContext":
+        self._session = self._exit_stack.enter_context(requests.Session())
+        self._session.headers.update(self._api_headers)
+        return self
 
-    Tracks user's permission for particular organizations/repositories.
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> Any:
+        self._session = None
+        self._exit_stack.close()
+
+    def _extract_auth(self, request: flask.Request) -> tuple[str | None, str]:
+        if request.authorization is None:
+            raise Unauthorized("Authorization required")
+
+        user = request.authorization.get("username")
+        token = request.authorization.password or request.authorization.token
+        if token is None:
+            _logger.warning(
+                f"Request to {self.org}/{self.repo} has no auth token"
+            )
+            raise Unauthorized("Authorization token required")
+        return user, token
+
+    def api_get(self, uri: str) -> dict[str, Any]:
+        if self._session is None:
+            raise RuntimeError(
+                "Calling CallContext.api_get() outside of a with block."
```

Comment:
`outside of a with block` gets the point across, but maybe you want to do something even more pointed, like "can only be used within a context manager defining a session".